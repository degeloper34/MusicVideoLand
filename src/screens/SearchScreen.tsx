import {FontAwesome} from "@expo/vector-icons";
import {useEffect, useState} from "react";
import {
  FlatList,
  Image,
  Pressable,
  StyleSheet,
  TextInput,
  View,
} from "react-native";
import {MusicVideo, MusicVideoList, RootTabScreenProps} from "../../types";
import {CustomText} from "../components/atoms";
import Colors from "../constants/Colors";
import {useAppSelector} from "../hooks/useRedux";
import musicVideoHelper from "../utils/musicVideoHelper";
import searchHelper from "../utils/searchHelper";

export default function SearchScreen({
  navigation,
}: RootTabScreenProps<"TabSearch">) {
  console.log("Render SearchScreen");
  const [searchedText, setSearchedText] = useState<string>("");
  const [filteredSearchedList, setFilteredSearchedList] =
    useState<MusicVideoList>([]);

  const {musicVideoList} = useAppSelector((state) => state?.musicVideoReducer);
  const {selectedGenreId} = useAppSelector((state) => state?.searchReducer);

  const onChangeSearchText = (text: string) => {
    setSearchedText(text);

    const searchedMusicVideoListByTitle = searchHelper.searchMusicVideoByTitle(
      musicVideoList,
      text
    );

    setFilteredSearchedList(searchedMusicVideoListByTitle);
  };

  const onPressFilter = () => {
    navigation.navigate("Modal");
  };

  useEffect(() => {
    console.log("useEffect SearchScreen");

    const musicVideoListByGenreId = musicVideoHelper.findMusicVideosByGenreId(
      musicVideoList,
      selectedGenreId
    );

    setFilteredSearchedList(musicVideoListByGenreId);
  }, [selectedGenreId]);

  const renderItem = ({item}: {item: MusicVideo}) => {
    return (
      <View style={styles.viewRenderItem}>
        <Image
          source={{uri: item.image_url}}
          resizeMode={"cover"}
          style={styles.imgMusicVideo}
          defaultSource={require("../assets/images/icon.png")}
        />
        <CustomText
          text={item?.title}
          type={"medium"}
          fontSize={14}
          numberOfLines={1}
        />
      </View>
    );
  };

  return (
    <View style={styles.container}>
      <View style={styles.viewSearchFilter}>
        <TextInput
          value={searchedText}
          placeholder="Search by title"
          style={styles.txtInputSearch}
          onChangeText={onChangeSearchText}
          autoFocus={true}
          returnKeyType={"done"}
          autoCapitalize="none"
          autoCorrect={false}
          clearButtonMode="always"
          maxLength={30}
        />

        <Pressable onPress={onPressFilter} style={styles.btnFilter}>
          <FontAwesome name={"filter"} color={Colors.black} size={20} />
        </Pressable>
      </View>

      <FlatList
        data={filteredSearchedList}
        renderItem={renderItem}
        contentContainerStyle={styles.flatListFilteredSearched}
        initialNumToRender={10}
        showsVerticalScrollIndicator={false}
        keyExtractor={(__, index) => String(index)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.black,
  },
  viewSearchFilter: {
    paddingHorizontal: 10,
    paddingTop: 20,
    flexDirection: "row",
  },
  txtInputSearch: {
    backgroundColor: Colors.white,
    borderRadius: 8,
    padding: 10,
    flex: 1,
    marginRight: 10,
  },
  btnFilter: {
    backgroundColor: Colors.white,
    padding: 10,
    borderRadius: 8,
  },
  flatListFilteredSearched: {
    paddingTop: 20,
    paddingHorizontal: 10,
  },
  imgMusicVideo: {
    width: 50,
    height: 50,
    borderRadius: 25,
    marginRight: 10,
  },
  viewRenderItem: {
    borderWidth: 1,
    borderColor: Colors.white,
    padding: 10,
    borderRadius: 8,
    marginBottom: 10,
    flexDirection: "row",
    alignItems: "center",
    overflow: "hidden",
  },
});
