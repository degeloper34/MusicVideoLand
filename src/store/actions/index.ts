export * from "./musicVideoActions";
export * from "./searchActions";
