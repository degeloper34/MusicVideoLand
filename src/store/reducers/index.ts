import {combineReducers} from "redux";

import musicVideoReducer from "./musicVideoReducers";
import searchReducer from "./searchReducers";

export const rootReducer = combineReducers({
  musicVideoReducer,
  searchReducer,
});
