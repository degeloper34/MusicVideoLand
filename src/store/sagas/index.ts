import {all} from "redux-saga/effects";

import {musicVideoWatcher} from "./musicVideoSagas";

export default function* rootSaga() {
  yield all([musicVideoWatcher()]);
}
