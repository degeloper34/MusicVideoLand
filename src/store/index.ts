export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

import {applyMiddleware, legacy_createStore as createStore} from "redux";
import createSagaMiddleware from "redux-saga";
import {rootReducer} from "./reducers";
import rootSaga from "./sagas";

const sagaMiddleware = createSagaMiddleware();

export const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

sagaMiddleware.run(rootSaga);
