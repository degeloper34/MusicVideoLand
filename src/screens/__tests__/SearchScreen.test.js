import renderer from "react-test-renderer";
import {render, fireEvent, waitFor} from "@testing-library/react-native";
import React from "react";
import SearchScreen from "../SearchScreen";
import {Provider} from "react-redux";
import {store} from "../../store";

describe("SearchScreen Tests", () => {
  test("SearchScreen | Render without issues", () => {
    const screen = renderer
      .create(
        <Provider store={store}>
          <SearchScreen />
        </Provider>
      )
      .toJSON();

    expect(screen).toMatchSnapshot();
  });

  test("SearchScreen | Navigate to ModalScreen without issues", async () => {
    const navigate = jest.fn();

    const {getByTestId} = render(
      <Provider store={store}>
        <SearchScreen navigation={{navigate}} />
      </Provider>
    );

    const toClick = await waitFor(() => getByTestId("btnFilter"));

    await fireEvent(toClick, "press");

    expect(navigate).toHaveBeenCalledWith("Modal");
  });
});
