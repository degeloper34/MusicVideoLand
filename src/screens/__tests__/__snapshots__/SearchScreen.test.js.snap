// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen Tests SearchScreen | Render without issues 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "paddingHorizontal": 10,
        "paddingTop": 20,
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      autoFocus={true}
      clearButtonMode="always"
      maxLength={30}
      onChangeText={[Function]}
      placeholder="Search by title..."
      returnKeyType="done"
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "flex": 1,
          "marginRight": 10,
          "padding": 10,
        }
      }
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 8,
          "padding": 10,
        }
      }
      testID="btnFilter"
    />
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingHorizontal": 10,
        "paddingTop": 20,
      }
    }
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    initialNumToRender={10}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</View>
`;
